swarm/pss/pss.go

```diff
@@ -241,91 +302,244 @@ func (self *Pss) checkFwdCache(addr []byte, digest pssDigest) bool {
+			log.Trace("attempting symmetric decrypt", "symkey", symkeyid)
+			symkey, err := self.w.GetSymKey(symkeyid)
+			if err == nil {
+				recvmsg, err = env.OpenSymmetric(symkey)
+				if err == nil {
+					if !recvmsg.Validate() {
+						return fmt.Errorf("validate sym enc msg fail")
+					}
+					from = self.reverseKeyPool[symkeyid][env.Topic]
+					log.Debug("pss message sym enc", "from", from, "payload", recvmsg.Payload)
+					break
+				}
+			} else {
+				log.Trace("sym decrypt failed", "symkey", symkeyid, "err", err)
+			}
+		}
+	}
+
+	// if we couldn't open the msg with any of the symkeys
+	// it should be an asymmetrically encrypted handshake sending a symkey to us
+	if recvmsg == nil {
+		recvmsg, err := env.OpenAsymmetric(self.privatekey)
+		// if it's not asym either, can't do anything with the msg. Drop it like it's hot
 		if err != nil {
-			return err
+			log.Debug("asym default decrypt of pss msg failed: %v", "err", err)
+			return nil
+		}
+		// check signature (if signed), strip padding
+		if !recvmsg.Validate() {
+			return fmt.Errorf("invalid message")
+		}
+		keymsgraw := recvmsg.Payload
+		keymsg := &pssKeyMsg{}
+		err = rlp.DecodeBytes(keymsgraw, keymsg)
+		if err == nil {
+			log.Trace("have symkeymsg", "from", keymsg.From)
+			copy(from[:], keymsg.From)
+			// TODO: need to handle / check for expired keys also here
+			_, err = self.SetOutgoingSymmetricKey(from, env.Topic, keymsg.Key)
+			if err != nil {
+				return fmt.Errorf("received invalid symkey in pss handshake for peer %x topic %x", keymsg.From, env.Topic)
+			}
+			// if we by now don't have keys for both in- and outgoing msgs, we need to make one for incoming and send it to the other party
+			// along with an encrypted secret so that it can tell that we received its key
+			// the encrypted secret will be our key encrypted with its key
+			if !self.isSecured(from, env.Topic) {
+				log.Trace("sending our symkey", "to", keymsg.From)
+				recvkeyid, err := self.GenerateIncomingSymmetricKey(from, env.Topic)
+				if err != nil {
+					return fmt.Errorf("set receive symkey fail (peer %x topic %x): %v", keymsg.From, env.Topic, err)
+				}
+				recvkey, err := self.w.GetSymKey(recvkeyid)
+				if err != nil {
+					return fmt.Errorf("get generated outgoing symkey fail (peer %x topic %x): %v", keymsg.From, env.Topic, err)
+				}
+				recvkeymsg := &pssKeyMsg{
+					From: self.BaseAddr(),
+					Key:  recvkey,
+				}
+				recvkeybytes, err := rlp.EncodeToBytes(recvkeymsg)
+				if err != nil {
+					return fmt.Errorf("rlp keymsg encode fail: %v", err)
+				}
+				// if the send fails it means this public key is not registered for this particular address AND topic
+				err = self.SendAsym(keymsg.From, env.Topic, recvkeybytes)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		// else we have successfully decoded the message and pass it to registered handlers
+	} else {
+		nid, _ := discover.HexID("0x00")
+		p := p2p.NewPeer(nid, fmt.Sprintf("%x", from), []p2p.Cap{})
```

Comment:
Given that both `nid` and `[]p2p.Cap{}` are effectively dummy values, does it not make sense just to be passing the `from` address to the handler rather than a dummy `p2p.Peer`?